doUpsert/index.js

```diff
@@ -27,37 +27,100 @@ module.exports = async function(context, jsonItem) {
-      await digitalTwin.upsertRelationship(
-          jsonItem.$sourceId,
-          jsonItem.$relationshipId,
-          jsonItem.relationship)
-          .catch((e) => {
-            context.log.error(`relationship ${jsonItem.$relationshipId}
-              on source ${jsonItem.$sourceId} insertion failed: `, e);
-            const err = `failed relationship: ${jsonString}`;
-            throw err;
-          });
-    })();
+    if (jsonItem.relationship.$relationshipDelete === true) { // relationship must be deleted
+      context.log.verbose(`deleting relationship ${jsonItem.$relationshipId}`);
+      await (async () => {
+        context.log.verbose('waiting 100ms');
+        sleep(100);
+        context.log.verbose('calling ADT relationship API');
+        await digitalTwin.deleteRelationship(
+            jsonItem.$sourceId,
+            jsonItem.$relationshipId)
+            .catch((e) => {
+              context.log.error(`relationship ${jsonItem.$relationshipId}
+                on source ${jsonItem.$sourceId} deletion failed: `, e);
+              const err = `failed relationship: ${jsonString}`;
+              throw err;
+            });
+      })();
+    } else { // relationship must be upserted
+      context.log.verbose(`upserting relationship ${jsonItem.$relationshipId}`);
+      await (async () => {
+        context.log.verbose('waiting 100ms');
+        sleep(100);
+        context.log.verbose('calling ADT relationship API');
+        await digitalTwin.upsertRelationship(
+            jsonItem.$sourceId,
+            jsonItem.$relationshipId,
+            jsonItem.relationship)
+            .catch((e) => {
+              context.log.error(`relationship ${jsonItem.$relationshipId}
+                on source ${jsonItem.$sourceId} insertion failed: `, e);
+              const err = `failed relationship: ${jsonString}`;
+              throw err;
+            });
+      })();
+    }
   } else if ('$id' in jsonItem) {
-    context.log.verbose(`upserting twin ${jsonItem.$id}`);
-    // twin
-    await (async () => {
-      context.log.verbose('waiting 20ms');
-      sleep(20);
-      context.log.verbose('calling ADT twin API');
-      await digitalTwin.upsertDigitalTwin(jsonItem.$id, jsonString)
-          .catch((e) => {
-            context.log.error(`twin ${jsonItem.$id} insertion failed: `, e);
-            const err = `failed twin: ${jsonString}`;
-            throw err;
-          });
-    })();
+    if (jsonItem.$entityDelete === true) { // twin must be deleted
+      // list all twin relationships
+      let rels = []
+      // outgoing relationships
+      let out = digitalTwin.listRelationships(jsonItem.$id);
+      for await (const rel of out) {
+        rels.push([rel.$sourceId, rel.$relationshipId])
+      }
+      // incoming relationships
+      let inc = digitalTwin.listIncomingRelationships(jsonItem.$id);
+      for await (const rel of inc) {
+        rels.push([rel.sourceId, rel.relationshipId])
```

Comment:
Another way to do in javascript is to use an object not an array, even for temporary vars, it creates clearer code, especially when using it (line 82 for example). Same for line 73.
`rels.push({ sourceId: rel.sourceId, id: rel.relationshipId })`

Line 82 example: `context.log.verbose(`deleting relationship ${tbd.id`);`